# Simple Machine Learning Code Demo with Scikit-Learn

In this notebook, you will find a practical step by step coding workflow using the Scikit-Learn library and its **California Housing** dataset. You will Learn about concepts like:
- Train and Test Split
- Baseline Model
- R2 Score
- Polynomial Features
- Hyperparameter Tuning
- and more...

If at any point, you find this notebook difficult to understand, please refer to my step by step video tutorial on YouTube: https://youtu.be/-IvNzmrcyUM

## Load Dataset

In the following cell, we will load the California Housing dataset and split it into features and targets. We will also print some basic information about the size and content of the dataset.

In [8]:
from sklearn import datasets

# load dataset from sklearn
housing = datasets.fetch_california_housing()

# split data to features and targets
x = housing.data
y = housing.target 

print("DATASET COLUMN NAMES:\n", housing.feature_names)
print("-----------------------")
print("nuber of rows:", len(x))
print("nuber of columns:", len(housing.feature_names))

DATASET COLUMN NAMES:
 ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']
-----------------------
nuber of rows: 20640
nuber of columns: 8


## Understand Sample and Target Data

In the cell below, we will print the first sample and the first target of the dataset, explaining what the stored information represents.
<br>
<br>
The information below represents a group of population who's:

- **Median Income**: $83,252
- **House Age**: 41 years old
- **Average Rooms**: 7 rooms (approx.)
- **Average Bedrooms**: 1 bedroom (approx.)
- **Population**: 322 people in this group.
- **Average Occup**: 2.5 occupants per household (approx.)
- **Latitude**: houses located in latitude 37.88.
- **Longitude**: houses located in longitude -122.23.

Finally, the **target** represents the dollar value of the house: $452,600

In [10]:
print(x[0]) # sample
print(y[0]) # target

[   8.3252       41.            6.98412698    1.02380952  322.
    2.55555556   37.88       -122.23      ]
4.526


## Train and Test Split

In the cell below, we will split our data into training samples and test samples.
<br>
It's an important step, when we reserve some of the data for model evaluation purposes. It will help us determine if the model actually learned the data or memorised it.
<br>
And the idea is, once the model is done training on the train set, we must expose it to samples it has never seen before (test set) to determine how well it performs. That's why we usually reserve about a 1/5th of the data for testing, instead of using the entire data during training.

In [23]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(
    x, 
    y, 
    test_size=0.2,
    random_state=432
)

print("number of train samples:", len(x_train))
print("number of test samples:", len(x_test))

number of train samples: 16512
number of test samples: 4128


## Train Model on Data Subset

In the cell below, we will choose a machine learning algorithm and we will train it on the California Housing Data 

In [24]:
from sklearn.linear_model import LinearRegression

# choose algorithm
model = LinearRegression()
# train model on the train set
model.fit(x_train, y_train)

LinearRegression()

## Evaluate Model on Data Subset
To understand how well-trained and intelligent our model is, we will need to test whether it learned or memorised the data (as mentioned above).
<br>
<br>
For this, we will make the model "guess" (or predict) the dollar value of samples it has never seen before (`x_test`). 
<br>
Once we get a prediction from our model, we'll compare it to the actual dollar value (given by `y_test`), which will produce a model score.
<br>
<br>
In our example, we will use the R2 metric to obtain a score, and since it's the first time we are evaluating this model, this score will represent our **Baseline**.

In [25]:
from sklearn.metrics import r2_score

# get the model to "guess" the dollar value of the houses in the test sample
y_pred = model.predict(x_test)
# compare what the model "guessed" with the actual value from y_test
r2 = r2_score(y_test, y_pred)

print("R2 Score:", r2)

R2 Score: 0.6080229586580348


## Optimization

Once we have a baseline model and a baseline R2 score, our goal is to improve it further and further. For example: At the moment, our score is `0.60`, which means that our model "understands" **60%** of why the house prices go up and down. The other **40%** is a mystery! Our objective in Machine Learning is to minimise the mystery as much as possible!
<br>
<br>
There are several techniques with which we can optimize this score, and you can find a few of them below.

### Optimize Features
The first optimization technique expands the number of features (or columns) and automatically generates more data for our model to rely on.
<br>
We will introduce Polynomial Features to our workflow, expanding our 8 existing features to 45 of them.
<br>
<br>
So in addition to the 8 features we currently have, we will add their squares (for example: 7 rooms and 49 rooms^2), and we will also multiply each column by another (for example: built year * number of occupants), resulting in 36 extra features!
<br>
Make sure to expand both the test data and the train data, as their number of columns must always match!

In [26]:
from sklearn.preprocessing import PolynomialFeatures

print("old numebr of features:", len(x_train[0]))

poly = PolynomialFeatures()
x_train = poly.fit_transform(x_train)
x_test = poly.fit_transform(x_test)

print("new numebr of features:", len(x_train[0]))

old numebr of features: 8
new numebr of features: 45


And now that we've expanded our data, let's try to train and evaluate our model once again, to see if we improved our baseline score:

In [27]:
model.fit(x_train, y_train) # train
y_pred = model.predict(x_test) # test
r2 = r2_score(y_test, y_pred) # evaluate

print("new R2 score:", r2)

new R2 score: 0.6610240202494676


**Congratulations!** Our new model is now 6 points smarter than the baseline model thanks to Polynomial Features!

### Optimize Algorithms
How do we know that LinearRegression is the best algorithm for our type of data? 
<br>
The answer is: **WE DON'T!**
<br>
Which means that we need to try several algorithms to determine that, in our case: 
LinearRegression, GradientBoostingRegresso, RandomForestRegressor.
<br>
Once we compair the R2 score for each model, we will know which algorithm works best!

In [28]:
from sklearn.ensemble import (
    GradientBoostingRegressor, 
    RandomForestRegressor
)

# initialize models
LR = LinearRegression()
GBR = GradientBoostingRegressor()
RFR = RandomForestRegressor()

# iterate over models, training and evaluating each
for model in [LR, GBR, RFR]:
    model.fit(x_train, y_train)
    y_pred = model.predict(x_test)
    r2 = r2_score(y_test, y_pred)
    print("MODEL:", model)
    print("R2 SCORE:", r2)
    print("-------------")

MODEL: LinearRegression()
R2 SCORE: 0.6610240202494676
-------------
MODEL: GradientBoostingRegressor()
R2 SCORE: 0.7919079837231457
-------------
MODEL: RandomForestRegressor()
R2 SCORE: 0.8049940006040666
-------------


We notice that the `RandomForestRegressor` and the `GradientBoostingRegressor` perform better than the baseline by more than 19 points!
<br>
Which means that they are a better fit than `LinearRegression`, and we should use one of them instead.
<br>
There's just one problem with these new algorithms - they take a relatively long time to process and they are very heavy on our computer. We will solve it in the cell below.

### Optimize Algorithm Speed
If we'd like to optimize the algorithm speed, rather than the R2 score, we can do the following:
- **RandomForestRegressor**: add `n_jobs=-1` argument to use the maximum available number of CPU cores for processing
- **GradientBoostingRegressor**: use optimized and more efficient version of this algorithm, `HistGradientBoostingRegressor`

Let's apply these in the cell below:

In [29]:
from sklearn.ensemble import HistGradientBoostingRegressor

# initialize speed optimized models
GBR = HistGradientBoostingRegressor()
RFR = RandomForestRegressor(
    n_jobs=-1
)

# iterate over models, training and evaluating each
for model in [GBR, RFR]:
    model.fit(x_train, y_train)
    y_pred = model.predict(x_test)
    r2 = r2_score(y_test, y_pred)
    print("MODEL:", model)
    print("R2 SCORE:", r2)
    print("-------------")

MODEL: HistGradientBoostingRegressor()
R2 SCORE: 0.8390043739506035
-------------
MODEL: RandomForestRegressor(n_jobs=-1)
R2 SCORE: 0.8045056753807682
-------------


We see that now, our optimized `GradientBoostingRegressor` model, produces a better R2 score than before! 
<br>
<br>
With `HistGradientBoostingRegressor`, we are now 4 points above the `GradientBoostingRegressor`score from the previous cell, and 24 points above the baseline! So let's choose this as our best-performing model and let's use it in the rest of this notebook.

### Optimize Hyperparameters
The last optimization technique we will see in this notebook, is optimization of hyperparameters for the best-performing model.
<br>
We can try different combinations of model parameter values to improve our score further and further, such as: `max_iter`, `learning_rate`, `max_depth`, etc.
<br>
<br>
In the cell below, we will find which number of trees works best for the `HistGradientBoostingRegressor`, optimizing a parameter named `n_features`

In [35]:
for i in [100, 200, 300, 400, 500]:
    model = HistGradientBoostingRegressor(
        max_iter = i
    )
    model.fit(x_train, y_train)
    y_pred = model.predict(x_test)
    r2 = r2_score(y_test, y_pred)
    print("NUMBER OF TREES:", i)
    print("R2 SCORE:", r2)
    print("-------------")

NUMBER OF TREES: 100
R2 SCORE: 0.8348898514124532
-------------
NUMBER OF TREES: 200
R2 SCORE: 0.8426991970214971
-------------
NUMBER OF TREES: 300
R2 SCORE: 0.8428739039341167
-------------
NUMBER OF TREES: 400
R2 SCORE: 0.847398290627727
-------------
NUMBER OF TREES: 500
R2 SCORE: 0.8461512103232809
-------------


We see that the best R2 score is given by `max_iter=400`, but that's not all! We can optimize more than one hyperparameter, so let's also introduce `learning_rate` into the miz with a nested for loop:

In [38]:
for j in [0.1, 0.05, 0.001]:
    for i in [200, 300, 400]:
        model = HistGradientBoostingRegressor(
            max_iter = i,
            learning_rate = j
        )
        model.fit(x_train, y_train)
        y_pred = model.predict(x_test)
        r2 = r2_score(y_test, y_pred)
        print("NUMBER OF TREES:", i)
        print("LEARNING RATE:", j)
        print("R2 SCORE:", r2)
        print("-------------")

NUMBER OF TREES: 200
LEARNING RATE: 0.1
R2 SCORE: 0.8415882755640507
-------------
NUMBER OF TREES: 300
LEARNING RATE: 0.1
R2 SCORE: 0.8458571678986928
-------------
NUMBER OF TREES: 400
LEARNING RATE: 0.1
R2 SCORE: 0.8418319639583227
-------------
NUMBER OF TREES: 200
LEARNING RATE: 0.05
R2 SCORE: 0.8388174987972418
-------------
NUMBER OF TREES: 300
LEARNING RATE: 0.05
R2 SCORE: 0.8440505516146202
-------------
NUMBER OF TREES: 400
LEARNING RATE: 0.05
R2 SCORE: 0.8466799096416513
-------------
NUMBER OF TREES: 200
LEARNING RATE: 0.001
R2 SCORE: 0.22127185476994615
-------------
NUMBER OF TREES: 300
LEARNING RATE: 0.001
R2 SCORE: 0.3033607401727726
-------------
NUMBER OF TREES: 400
LEARNING RATE: 0.001
R2 SCORE: 0.3734921617612692
-------------


We see that the best R2 score is `0.846` given by:
- `max_iter=400`
- `learning_rate=0.05`

So we should set it as our new best-performing model.
<br>
<br>
Please note, we've experimented only with 2 hyperparameters in this workflow, but in reality - you should experiment with as many hyperparameters combinations as possible!

## Model Saving
Great! So we found our best performing model, which is almost 25 points better than the baseline! Next, we will need to save this model to our system, so we never have to train it again.
<br>
<br>
For this we will use a library named `joblib`:

In [39]:
import joblib

model = HistGradientBoostingRegressor(
    max_iter=400, # best max iter
    learning_rate=0.05 # best learning rate
)
model.fit(x_train, y_train)

# save model
joblib.dump(model, "my_model.joblib")

['my_model.joblib']

### Ensure Model is Properly Saved
To make sure our model was properly saved, we will load (also with joblib) it and compare it to the existing model.

In [40]:
# evaluate existing model
y_pred = model.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("existing model score:", r2)

# load saved model
saved_model = joblib.load("my_model.joblib")
# evaluate saved model
y_pred = saved_model.predict(x_test)
r2 = r2_score(y_test, y_pred)
print("saved model score:", r2)

existing model score: 0.8409061745888814
saved model score: 0.8409061745888814


If the scores of both models are a perfect match to the smallest of decimal points, as above, our model was successfully saved!

## Further Development Ideas
- try optimizing more than 2 hyperparameters, reaching scores above 86%
- instead of simply printing the R2 scores, try visualising them with Pandas or Matplotlib
- Find even better algorithms that outperform `HistGradientBoostingRegressor`
- Find a different dataset and analyze it like we've analyzed this one.